src/tsd/RpcPluginsManager.java

```diff
@@ -0,0 +1,593 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Pattern;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.Atomics;
+import com.stumbleupon.async.Callback;
+import com.stumbleupon.async.Deferred;
+
+import org.jboss.netty.channel.Channel;
+import org.jboss.netty.handler.codec.http.HttpMethod;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.opentsdb.BuildData;
+import net.opentsdb.core.Aggregators;
+import net.opentsdb.core.TSDB;
+import net.opentsdb.stats.StatsCollector;
+import net.opentsdb.utils.JSON;
+import net.opentsdb.utils.PluginLoader;
+
+/**
+ * Manager for the lifecycle of <code>HttpRpc</code>s, <code>TelnetRpc</code>s,
+ * <code>RpcPlugin</code>s, and <code>HttpRpcPlugin</code>.  This is intended
+ * to be a singleton.
+ * 
+ * @since 2.1
+ */
+public final class RpcPluginsManager {
+  private static final Logger LOG = LoggerFactory.getLogger(RpcPluginsManager.class);
+  
+  @VisibleForTesting
+  protected static final String PLUGIN_BASE_WEBPATH = "plugin";
+  
+  /** Matches incoming HTTP requests if they are for HTTP RPC plugins. */
+  private static final Pattern PLUGIN_WEBPATH_FOR_REQUESTS = Pattern.compile(
+      "^/?" + PLUGIN_BASE_WEBPATH + "/.+", 
+      Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
+  
+  /** Matches paths declared by {@link HttpRpcPlugin}s that are rooted in
+   * the system's plugins path. */
+  private static final Pattern PAT = Pattern.compile(
+      "^/?" + PLUGIN_BASE_WEBPATH + "/?.*", 
+      Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
+  
+  /** Reference to our singleton instance.  Set in {@link #initialize}. */
+  private static final AtomicReference<RpcPluginsManager> INSTANCE = Atomics.newReference();
+
+  /** Commands we can serve on the simple, telnet-style RPC interface. */
+  private ImmutableMap<String, TelnetRpc> telnet_commands;
+  /** Commands we serve on the HTTP interface. */
+  private ImmutableMap<String, HttpRpc> http_commands;
+  /** HTTP commands from user plugins. */
+  private ImmutableMap<String, HttpRpcPlugin> http_plugin_commands;
+  /** List of activated RPC plugins */
+  private ImmutableList<RpcPlugin> rpc_plugins = null;
+  
+  /** The TSDB that owns us. */
+  private TSDB tsdb;
+  
+  public RpcPluginsManager() {
+    // Default values before initialize is called.
+    telnet_commands = ImmutableMap.of();
+    http_commands = ImmutableMap.of();
+    http_plugin_commands = ImmutableMap.of();
+    rpc_plugins = ImmutableList.of();
+  }
+  
+  /**
+   * @param tsdb The parent TSDB object
+   */
+  public void initialize(final TSDB tsdb) {
```

Comment:
We may be able to skip the initialization call here and just perform the init in the constructor since we should only ever have one instance of the rpc manager.